transports/dns/src/lib.rs

```diff
@@ -107,134 +171,150 @@ where
-                            Ok(list) => Ok(list.map(|s| s.ip()).collect::<Vec<_>>()),
-                            Err(e) => Err(e),
-                        });
-                    });
-
-                    let (dns4, dns6) = match cmp {
-                        Protocol::Dns(_) => (true, true),
-                        Protocol::Dns4(_) => (true, false),
-                        Protocol::Dns6(_) => (false, true),
-                        _ => unreachable!(),
+        // Asynchronlously resolve all DNS names in the address before proceeding
+        // with dialing on the underlying transport.
+        Ok(async move {
+            let resolver = self.resolver;
+            let inner = self.inner;
+
+            let mut last_err = None;
+            let mut attempts = 0;
+            let mut dnsaddr_lookups = 0;
+            // We optimise for the common case of a single DNS component
+            // in the address that is resolved with a single lookup.
+            let mut unresolved = SmallVec::<[Multiaddr; 1]>::new();
+            unresolved.push(addr.clone());
+
+            // Resolve (i.e. replace) all DNS protocol components, initiating
+            // dialing attempts as soon as there is another fully resolved
+            // address.
+            while let Some(addr) = unresolved.pop() {
+                if let Some((i, name)) = addr.iter().enumerate().find(|(_, p)| match p {
+                    Protocol::Dns(_) |
+                    Protocol::Dns4(_) |
+                    Protocol::Dns6(_) |
+                    Protocol::Dnsaddr(_) => true,
+                    _ => false
+                }) {
+                    match resolve(&name, &resolver).await {
+                        Err(e) => {
+                            if unresolved.is_empty() {
+                                return Err(e)
+                            }
+                            // If there are still unresolved addresses, there is
+                            // a chance of success, but we track the last error.
+                            last_err = Some(e);
+                        }
+                        Ok(Resolved::One(ip)) => {
+                            log::trace!("Resolved {} -> {}", name, ip);
+                            let addr = addr.replace(i, |_| Some(ip)).expect("`i` is a valid index");
+                            unresolved.push(addr);
+                        }
+                        Ok(Resolved::Many(ips)) => {
+                            for ip in ips {
+                                log::trace!("Resolved {} -> {}", name, ip);
+                                let addr = addr.replace(i, |_| Some(ip)).expect("`i` is a valid index");
+                                unresolved.push(addr);
+                            }
+                        }
+                        Ok(Resolved::Addrs(addrs)) => {
+                            if dnsaddr_lookups < MAX_DNSADDR_LOOKUPS {
+                                dnsaddr_lookups += 1;
```

Comment:
Say that a single `/dnsaddr` record contains a very large amount of bogus `dns6` addresses. Given that `dnsaddr_lookups` is only increased on successful `dnsaddr` lookups, the node would try to resolve each bogus `dns6` address possibly consuming a lot of resources.

Would it not be safer to keep count of the overall number of calls to `resolve()` and terminate the loop after some constant X?

Reply:
> Say that a single /dnsaddr record contains a very large amount of bogus dns6 addresses.

I assume you mean when a single `/dnsaddr/...` lookup returns a lot of such TXT records. A single TXT record only contains a single address as per [the spec](https://github.com/multiformats/multiaddr/blob/master/protocols/DNSADDR.md) and thus we only process the first `<character-string>` of each TXT record, which can only be up to 255 bytes in length. There can be many TXT records in a reply, of course, but then there are also the size limits on DNS responses (anything > 512 bytes is already undesirable because, in the absence of [EDNS(0)](https://en.wikipedia.org/wiki/Extension_mechanisms_for_DNS) one has too fall back to TCP).

> Given that dnsaddr_lookups is only increased on successful dnsaddr lookups, the node would try to resolve each bogus dns6 address possibly consuming a lot of resources.

Not sure about whether that would consume a lot of resources, but I think counting the `resolve()` calls would be fine. Even beyond that, it may be preferable if we have a fixed limit on the number of TXT records that we process from a `/dnsaddr` lookup response, just to be on the safe side.

Reply:
What do you think about https://github.com/libp2p/rust-libp2p/pull/1931/commits/5b99f128d412fb805929d4853bd7c85dee3450a5?

Reply:
A small further refinement for better troubleshooting in case one runs into the limits: https://github.com/libp2p/rust-libp2p/pull/1931/commits/4e8d8f2d311808b82ca32f8d520f22efb3a99a8d.

Reply:
> I assume you mean when a single /dnsaddr/... lookup returns a lot of such TXT records.

Yes, sorry for the confusion and thanks for the detailed reply.

> There can be many TXT records in a reply, of course, but then there are also the size limits on DNS responses (anything 512 bytes is already undesirable because, in the absence of EDNS(0) one has too fall back to TCP).

More out of curiosity, even if the DNS server has to fall back to TCP to deliver the large DNS response, that does not defend against the attack of the response containing a large amount of TXT records, right? If I understand correctly trust-dns handles both UDP and TCP by default.

Limit enforcing via `MAX_DNS_LOOKUPS` and `MAX_TXT_RECORDS` look good to me.

Reply:
> More out of curiosity, even if the DNS server has to fall back to TCP to deliver the large DNS response, that does not defend against the attack of the response containing a large amount of TXT records, right? If I understand correctly trust-dns handles both UDP and TCP by default.

Yes, that is correct. It is just that due to the "expensive" TCP fallback, "honest" DNS responses rather tend to be small to avoid the TCP fallback, whilst for an attack returning large responses is probably a lot less interesting than returning false records, i.e. performing DNS spoofing / cache poisoning.

> Limit enforcing via MAX_DNS_LOOKUPS and MAX_TXT_RECORDS look good to me.

:+1: Thanks for reviewing.